lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +434,1140 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+      else
+        Result.insert({ CI, FunctionEdgeTypeValue::LongJmp });
+    }
+  }
+
+  bool FoundRet, FoundBrokenRet;
+  FoundRet = FoundBrokenRet = false;
+  for (const auto &E : Result) {
+    if (E.second == FunctionEdgeTypeValue::Return)
+      FoundRet = true;
+    if (E.second == FunctionEdgeTypeValue::BrokenReturn)
+      FoundBrokenRet = true;
+  }
+
+  FunctionTypeValue Type;
+  if (FoundRet) {
+    Type = FunctionTypeValue::Regular;
+  } else if (FoundBrokenRet) {
+    Type = FunctionTypeValue::Fake;
+  } else {
+#ifdef NORETURN_IS_FIXED
+    Type = FunctionTypeValue::NoReturn;
+#else
+    Type = FunctionTypeValue::Regular;
+#endif
+  }
+
+  // Finally retrieve the clobbered registers and return them.
+  std::erase_if(ClobberedRegs, [&](const auto &E) {
+    return CalleeSavedRegs.find(E) != CalleeSavedRegs.end();
+  });
+
+  // Finalize CFG for the model.
+  for (const auto &[CI, EdgeValue] : Result) {
+    auto PC = MetaAddress::fromConstant(CI->getArgOperand(2));
+    auto &Block = CFG.at(PC);
+    Block.Successors.insert(makeEdge(MetaAddress::invalid(), EdgeValue));
+  }
+
+  // We say that a register is callee-saved when, besides being preserved by the
+  // callee, there is at least a write onto this register.
+  std::set<GlobalVariable *> ActualCalleeSavedRegs;
+  std::set_intersection(CalleeSavedRegs.begin(),
+                        CalleeSavedRegs.end(),
+                        WrittenRegisters.begin(),
+                        WrittenRegisters.end(),
+                        std::inserter(ActualCalleeSavedRegs,
+                                      ActualCalleeSavedRegs.begin()));
+
+  // Union between effective callee-saved registers and SP.
+  std::set<GlobalVariable *> CalleeSavedAndStackPointerRegs = {
+    ActualCalleeSavedRegs
+  };
+  CalleeSavedAndStackPointerRegs.insert(GCBI->spReg());
+
+  // Refine ABI analyses by suppressing callee-saved and stack pointer registers
+  // from the final results.
+  for (const auto &Reg : CalleeSavedAndStackPointerRegs) {
+    if (ABIResults.Arguments.count(Reg) != 0)
+      ABIResults.Arguments[Reg] = model::RegisterState::Values::No;
+  }
+
+  for (const auto &[K, _] : ABIResults.CallSites) {
+    for (const auto &Reg : CalleeSavedAndStackPointerRegs) {
+      if (ABIResults.CallSites[K].Arguments.count(Reg) != 0)
+        ABIResults.CallSites[K]
+          .Arguments[Reg] = model::RegisterState::Values::No;
+
+      if (ABIResults.CallSites[K].ReturnValues.count(Reg) != 0)
+        ABIResults.CallSites[K]
+          .ReturnValues[Reg] = model::RegisterState::Values::No;
+    }
+  }
+
+  for (const auto &E : Result) {
+    if (E.second != FunctionEdgeTypeValue::Return
+        && E.second != FunctionEdgeTypeValue::IndirectTailCall) {
+      auto PC = MetaAddress::fromConstant(E.first->getOperand(2));
+
+      auto It = ABIResults.ReturnValues.find(PC);
+      if (It != ABIResults.ReturnValues.end())
+        ABIResults.ReturnValues.erase(PC);
```

Comment:
What's going on here? Why are we doing this?
Document.

Reply:
We are not keeping the results of `BrokenReturn`s, right?

Reply:
Elaborated.